### ISSUE TITLE
rnn training example?

### ISSUE THREAD

**Issue by mingfeima** [intel/mkl-dnn#218]
is there going to be a rnn training example? say how to allocate the `workspace` shared between rnn forward and rnn backward?

**Comment by emfomenk** [comment]
Workspace allocation and sharing between forward and backward passes are common for the entire MKL-DNN API. The typical steps are:
1. create forward primitive descriptor (note that `prop_kind` should be equal to `forward_training`)
2. query this primitive descriptor for workspace memory primitive descriptor
3. if that workspace_pd is not NULL, query its size via ` mkldnn_memory_primitive_desc_get_size()`
4. allocate enough memory for the workspace, create workspace memory and set handle to allocated memory
5. pass workspace as one of the outputs for forward primitive
6. on backward pass put the same workspace as one of the inputs

The corresponding places can be found:
- C example: [lines 436 .. 450](https://github.com/intel/mkl-dnn/blob/master/examples/simple_training_net.c#L436)
- C++ example: [lines 215 .. 220](https://github.com/intel/mkl-dnn/blob/master/examples/simple_training_net.cpp#L215)

**Comment by emfomenk** [comment]
HI @mingfeima,

GRU is something we are currently working on. Do you have any comments in what particular implementation you are interested in? There are a bunch of different incompatible implementations (TF, MxNet, PaddlePaddle, cuDNN, ...).

Alas, dropout is not supported. Could you please elaborate on where is it used? Or the ask is more about compatibility?

**Comment by mingfeima** [comment]
@emfomenk i am working on PyTorch in which GRU is implemented in the following way, should be identical to cuDNN version. I wasn't aware of other implementations so i can't make compare.
```python
    gi = F.linear(input, w_ih, b_ih)
    gh = F.linear(hidden, w_hh, b_hh)
    i_r, i_i, i_n = gi.chunk(3, 1)
    h_r, h_i, h_n = gh.chunk(3, 1)

    resetgate = F.sigmoid(i_r + h_r)
    inputgate = F.sigmoid(i_i + h_i)
    newgate = F.tanh(i_n + resetgate * h_n)
    hy = newgate + inputgate * (hidden - newgate)
```

And yes, it's more about compatibility at the moment from my side.

**Comment by mratsim** [comment]
I've just started implementing GRU in my own deep learning framework and did a comprehensive overview of all the implementations in the Wild [here](https://github.com/mratsim/Arraymancer/issues/228), including a reference to the heavy optimizations proposed by Baidu Research.

Here are the equations I've found:
- `◦` is element-wise multiplication
- `*` is tensor multiplication
    Tensorflow
    ```
    r = sigmoid(W_{ir} * x + b_{ir} + W_{hr} * h + b_{hr}) 
    z = sigmoid(W_{iz} * x + b_{iz} + W_{hz} * h + b_{hz})
    n = tanh(W_{in} * x + b_{in} +  W_{hn} * (r ◦ h) + b_{hn}))
    h' = (1 - z) ◦ n + z ◦ h
    ```
    PyTorch equations
    ```
    r = sigmoid(W_{ir} * x + b_{ir} + W_{hr} * h + b_{hr}) 
    z = sigmoid(W_{iz} * x + b_{iz} + W_{hz} * h + b_{hz})
    n = tanh(W_{in} * x + b_{in} + r ◦ (W_{hn} * h + b_{hn}))
    h' = (1 - z) ◦ n + z ◦ h
    ```
    Note that in the paper equations are:
    ```
    r = sigmoid(W_{ir} * x + b_{ir} + W_{hr} * h + b_{hr}) 
    z = sigmoid(W_{iz} * x + b_{iz} + W_{hz} * h + b_{hz})
    n = tanh(W_{in} * x + b_{in} +  W_{hn} * (r ◦ h) + b_{hn}))
    h' = (1 - z) ◦ h + z ◦ n
    ```
    And CuDNN
    ```
    it = σ(Wi * xt + Ri * ht-1 + bWi + bRu)
    rt = σ(Wr * xt + Rr * ht-1 + bWr + bRr)
    h't = tanh(Wh * xt + rt ◦ (Rh * ht-1 + bRh) + bWh)
    ht = (1 - it)◦h't + it◦ht-1
    ```

Note that PyTorch/CuDNN have the same equations which allows reusing weights. Reusing weights between CPU and GPU is a pain point in Keras at the moment:
  - [Keras: weights on GPU cannot be reused on CPU](https://github.com/keras-team/keras/issues/8860) and solutions (i.e. redoing a CPU layer):
    - [LSTM](https://github.com/keras-team/keras/pull/8307)
    - [GRU](https://github.com/keras-team/keras/pull/9112)

My own implementation in [Nim](https://nim-lang.org/) outputs the same result as PyTorch but limits the number of temporary allocations by reusing buffers and leveraging loopfusion (`sigmoid(a+b)` or `tanh(x + y ◦ z)` in single pass.

```Nim
proc gru_cell_forward*[T: SomeReal](input, hidden,
                                    w_input, w_recur,
                                    b_input, b_recur: Tensor[T],
                                    result: var Tensor[T]) =
  let
    H = hidden.shape[1]
    # Slices
    sr = (0 ..< H)|1
    sz = (H ..< 2*H)|1
    srz = (0 ..< 2*H)|1
    sn = (2*H ..< 3*H)|1

  var Wx, Rh: Tensor[T] # TODO, pass those as parameter to allow buffer reuse
  # Resulting shape [batch_size, 3*H]
  linear(input, w_input, b_input, Wx)
  linear(hidden, w_recur, b_recur, Rh)

  # To reduce allocations, we compute reset gate r
  # and update gate z in the previous buffers
  # We keep them concatenated to improve throughput
  var reset_update = Wx[_, srz] # shape [batch_size, 2*H]
  apply2_inline(reset_update, Rh[_, srz]):
    sigmoid(x + y)

  # We also reuse the previous buffer for the candidate hidden state n
  # shape [H, batch_size]
  var n = Wx[_, sn] # shape [batch_size, H]
  apply3_inline(n, reset_update[_, sr], Rh[_, sn]):
    tanh(x + y * z)

  # Compute the next hidden state
  result = map3_inline(Wx[_, sz], n, hidden):
    (1 - x) * y + x * z
```

**Comment by vpirogov** [comment]
@mratsim, thanks for sharing your findings and insights on the weight reuse problem. @igorsafo, @irinasok is this consistent with the picture we have on GRU flavors?

**Comment by irinasok** [comment]
Hi @mingfeima, @mratsim

**Comment by vpirogov** [comment]
Initial GRU support is added in 0282e6b26fae3be47aad6996a611cbcf99b29949. Let us know what you think.

**Comment by mingfeima** [comment]
@vpirogov as long as cuDNN flavor of GRU is provided, i am fine with current interface.
@emfomenk got a few questions when integrating MKLDNN RNN into pytorch.

1. how MKLDNN differs `bias_ih` and `bias_hh`? and also `grad_bias_ih` and `grad_bias_hh`? In the [simple_rnn](https://github.com/intel/mkl-dnn/blob/master/examples/simple_rnn.cpp) example, there is only one bias described by `mkldnn::memory::format::ldgo`
2. which is the memory::format for `hx`, `cx`, `hy`, `cy`,the following is pytorch shape
`hx` - hidden state input, shape _(num_layers * num_directions, batch, hidden_size)_
`cx` - cell state input, shape _(num_layers * num_directions, batch, hidden_size)_
`hy` - hidden state output, shape _(num_layers * num_directions, batch, hidden_size)_
`cy` - cell state output, shape _(num_layers * num_directions, batch, hidden_size)_
is it [mkldnn_ldsnc](https://github.com/intel/mkl-dnn/blob/72236df2da36d2c73ce1c719c6e9fa04a2f6bdc7/include/mkldnn_types.h#L307)? if so, `num_states` refers to `hx` and `cx` for LSTM, `hx` for GRU?
then i need to cancat `hx` and `cx` into a 5D tensor along `num_states`?
3. Do you implement layer fusion? basically i need to know if there is performance penalty in case i break a multi-layer RNN primitive into separate layer primitives. 
4. Do you have [packed_sequence](https://pytorch.org/docs/stable/_modules/torch/nn/utils/rnn.html) input at the moment? Right now i am unpacking the packed input myself and shipping input to MKLDNN in TNC format.

**Comment by mingfeima** [comment]
@mgouicem thanks for the update!

1. So for the backward path, i should give `(bias_ih + bias_hh)/2` to both `bias_ih` and `bias_hh`? i suppose mathematically, this should not cause any trouble.
2. Naively, `hx` and `cx` should be concat once for forward and once for backward. Anyway, i guess i can do the same trick here as i did on the layer weight tensors.
3. have you got any code example or documentation for the usage of `view` in packed sequences? So from the performance perspective, it should be beneficial to treat unpacked sequences in the same manner with packed ones?

**Comment by lihaofd** [comment]
@mgouicem We are trying to integrate MKLDNN rnn into mxnet for testing, is it possible to provide a sample like one layer unidirect lstm forward/backward with des/memory creation of input and output like hx/cx/x, wx/wh/bx/bh, y/hy etc?

**Comment by lihaofd** [comment]
I am trying to test one layer lstm unidirect forward. https://github.com/lihaofd/incubator-mxnet/blob/mkldnn-rnn/src/operator/nn/mkldnn/mkldnn_rnn_impl.h

void MKLDNNLstmForwardInferenceSingleLayer()

Many of codes are based on guessing. So far the result of y/hy can't match with cudnn's result when using same input.
1. Is there any code incorrect related to input: x, wx/wh/bx/bh and output: y/hy?
2, how to deal with input like hx, cx?
3. how to deal with output like cy?

Thanks!

**Comment by mgouicem** [comment]
@mingfeima 
1. Mathematically, the sum on forward translates to a copy on bwd. So if `bias = bias_hh + bias ih` then `grad_bias_hh = grad_bias` and `grad_bias_ih = bias_ih`
2. Ok, good to know. Out of curiosity, what is the trick you use to avoid the concat? :-)
3. We will make a full and comprehensive example available to showcase the unpacked sequence handling. In the case of the packet sequence, all MKL-DNN memories are "dense", so for `src_layer` and `dst_layer`, you will have to create a memory per chunk, and set the data handle appropriately. For `src_iter / dst_iter`, you can use view to pass them from chunk to chunk, or to only update the appropriate part of `dst_iter`.
From a performance point of view, it should not matter whether you use packed or unpacked sequences. Just use whatever you have to avoid copies, and that is semantically equivalent to the framework op.
4. This is an example where we use a view on a memory to avoid spurious copies. There is no obligation to do this of course, you can use separate memories for `src_iter` and `dst_iter`. In the context of this GNMT decoder, it will just introduce spurious copies.

@lihaofd we have an example that showcases forward inference in examples/simple_rnn.cpp. We will make a simple training example available too.

**Comment by lihaofd** [comment]
@mgouicem, I collected code from examples/simple_rnn.cpp to create MKLDNNLstmForwardInferenceSingleLayer, it still doesn't match the cudnn result in just one layer forward...
Any suggestions on it? thanks

**Comment by mgouicem** [comment]
@lihaofd just gave a quick look. It seems that you are using a wrong layout for weights (lines 90 - 96).
In your code, `weights_layer` is of shape ldigo -- (1, 1, I, 4, H) in your notation-- and `weights_iter` is of shape ldigo too -- (1, 1, H, 4, H) with your notations. So according to line 46, you either need to reorder your weights, or use ldgoi if your weights already have the right format.

Also note that we expect a certain order for the gate weights. Using the formulas in [MXNet documentation ](https://mxnet.incubator.apache.org/api/python/gluon/rnn.html#mxnet.gluon.rnn.LSTM), here are the correspondance with MKL-DNN :
- we have only one bias, so we take the sum of the biases for each gate
- gate0 is the `f` gate
- gate1 is the `i` gate
- gate2 is the `o` gate
- gate3 is the `g` gate

**Comment by mgouicem** [comment]
@lihaofd 
Some more answers:

> 1. Is there any code incorrect related to input: x, wx/wh/bx/bh and output: y/hy?

Could you please clarify the question?

> 2. how to deal with input like `hx, cx`?

`hx` and `cx` map to `src_iter`. Because we use one tensor to group the two (with `ldsnc` layout), you have to concatenate them w.r.t the `s` dimension of the `ldsnc` layout.

> 3. how to deal with output like `cy`?

`hy` and `cy` map to `dst_iter`. By the same argument as 2., you will have to split dst_iter on the s dimension to produce `hy` and `cy`.

**Comment by lihaofd** [comment]
@mgouicem  thank you very much for your help!
Is there any code incorrect related to input: x, wx/wh/bx/bh and output: y/hy?
Could you please clarify the question?

My question is I doubt there is some code incorrect when I transfer the input/output data and weight into mkldnn desc/memory etc. It seems like there has issue relate to my weight's layout and gate order.
I am fixing my code and will get back to you later. thanks again.

**Comment by lihaofd** [comment]
@mgouicem after fixing the code, the testing shows unidirect LSTM inference can match the cudnn result.
But when I trying mulitiple layers bidirect LSTM inference, it goes to crash (single layer bidirect inference can work)
https://github.com/lihaofd/incubator-mxnet/blob/mkldnn-rnn/src/operator/nn/mkldnn/mkldnn_rnn_impl.h
The code crashed on line 439 to 444. 
I am testing L = 3 layers bidirect (bidirectional_concat) LSTM, the layer 2 and 3 have same inputsize and hidden size, so I put these 2 layers together into rnn_forward::desc layer_desc on Line 439 rnn_forward::desc layer_desc

T: Timestep, N: Batchsize, I: Inputsize, H: Hidden size, D =2 (bidirectional_concat)
memory::dims src_layer_tz = {T, N, I};
    memory::dims weights_layer_tz = {L - 1, D, I, 4, H};  //  ldigo
    memory::dims weights_iter_tz = {L - 1, D, H, 4, H};  //  ldigo
    memory::dims bias_tz = {L - 1, D, 4, H};   //ldgo
    memory::dims dst_layer_tz = {T, N, D * H};   //tnc
    memory::dims src_iter_tz = {L - 1, D, 2, N, H};  //  ldsnc
    memory::dims dst_iter_tz = {L - 1, D, 2, N, H};  //  ldsnc
While it failed on line 443 - 444.
could you help to check if there is anything wrong in layout or anything else?

Thanks!

**Comment by mgouicem** [comment]
@lihaofd could you send the specific error you are getting?
From the code, I can see a problem (do not know if it is related to the error you are seeing though).
If you have `I=H`, `D=2` and `L>2`, you will run the first layer as bidirectional and `memory::dims dst_layer_tz_0 = {T, N, D * H};`. However, this will be the input of the next layer, but for the next layer you have `memory::dims src_layer_tz = {T, N, I};` and these do not match.
 
My recommendation would be to not split the layers and fuse them all if possible.
If not possible, you will have to separately run the left2right and right2left passes, and concatenate their result (and if you want to optimize, you can avoid this concatenation with views on the same memory)

**Comment by lihaofd** [comment]
@mgouicem , the first layer input is (T, N, I), output is (T, N, D * H), from second layer "I" has been set as  I= D * H in line 275, so the input memory::dims src_layer_tz = {T, N, I} is = (T, N, D * H)
The failed error on  auto prim_desc
        = mkldnn::rnn_forward::primitive_desc(layer_desc, cpu_engine); it shows
terminate called after throwing an instance of 'dmlc::Error'
  what():  [06:38:13] src/engine/./threaded_engine.h:379: std::exception
A fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.

Stack trace returned 8 entries:
[bt] (0) /home/hli32/mxnet/mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::StackTrace()+0x3f) [0x7f5c367c7b5f]
[bt] (1) /home/hli32/mxnet/mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x21) [0x7f5c367c7f31]
[bt] (2) /home/hli32/mxnet/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*)+0xa75) [0x7f5c393687f5]
[bt] (3) /home/hli32/mxnet/mxnet/python/mxnet/../../lib/libmxnet.so(std::_Function_handler<void (std::shared_ptr<dmlc::ManualEvent>), mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::{lambda()#1}::operator()() const::{lambda(std::shared_ptr<dmlc::ManualEvent>)#1}>::_M_invoke(std::_Any_data const&, std::shared_ptr<dmlc::ManualEvent>)+0x92) [0x7f5c39378522]
[bt] (4) /home/hli32/mxnet/mxnet/python/mxnet/../../lib/libmxnet.so(std::thread::_Impl<std::_Bind_simple<std::function<void (std::shared_ptr<dmlc::ManualEvent>)> (std::shared_ptr<dmlc::ManualEvent>)> >::_M_run()+0x44) [0x7f5c39367c54]
[bt] (5) /usr/lib64/libstdc++.so.6(+0xb52b0) [0x7f5cbce402b0]
[bt] (6) /usr/lib64/libpthread.so.0(+0x7e25) [0x7f5cd5d54e

**Comment by mgouicem** [comment]
@lihaofd By looking at [MXNet code for _RNNLayer](https://mxnet.incubator.apache.org/_modules/mxnet/gluon/rnn/rnn_layer.html), it seems that the bidirectional semantics is different than what you implemented.

In your case, because you split the first layer, you end up with weights_layer of size `{1, D, I, 4, H}` for the first layer, and `{L-1, D, D*H, 4, H}` for the `L-1` remaining layers (the `D*H` here is likely the cause of the segfault). However, MXNet semantics suggests that you need `{L, D, I, 4, H}` for the L layers. If we assume that `I=H`, here you may be referencing `(L-1)*D*4*H` extra weights.

So I would suggest you to try with:
`memory::dims weights_layer_all_tz = {L, D, I, 4, H}`
`std::vector user_wei_layer_all(L*D*I*4*H)`

**Comment by vpirogov** [comment]
RNN training examples is [now available](https://github.com/intel/mkl-dnn/blob/master/examples/simple_rnn_training.cpp). With that I'm closing the issue. If you have further questions feel free to submit a new one or reopen.
